### ISSUE TITLE
Crash in View.of when accessed through ScrollPosition.devicePixelRatio

### ISSUE THREAD

**Issue by Piinks** [flutter/flutter#126454]
b/279930080

Customer reports a crash in production that is coming from View.of being called through ScrollPosition.devicePixelRatio.

It looks like the NestedScrollView is involved as well. 

ScrollPosition.devicePixelRatio recently changed in #117916, @goderbauer FYI

```Thread 0 main thread_TypeError: widgets librarycontent_copy
at ??View.of(view.dart:107)
at ??ScrollPosition.devicePixelRatio(scroll_position.dart:244)
at ??ScrollMetrics.copyWith(scroll_metrics.dart:57)
at ??ScrollPosition.didUpdateScrollDirection(scroll_position.dart:922)
at ??_NestedScrollCoordinator.updateUserScrollDirection(nested_scroll_view.dart:634)
at ??_NestedScrollPosition.goIdle(nested_scroll_view.dart:1344)
at ??new _NestedScrollPosition(nested_scroll_view.dart:1184)
at ??_NestedScrollController.createScrollPosition(nested_scroll_view.dart:1119)
at ??ScrollableState._updatePosition(scrollable.dart:565)
at ??ScrollableState.didChangeDependencies(scrollable.dart:599)
at ??StatefulElement._firstBuild(framework.dart:5237)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement._firstBuild(framework.dart:5242)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement._firstBuild(framework.dart:5242)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at 
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ComponentElement._firstBuild(framework.dart:5068)
at ??ComponentElement.mount(framework.dart:5062)
at ??Element.inflateWidget(framework.dart:3971)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??RenderObjectElement.updateChildren(framework.dart:6093)
at ??MultiChildRenderObjectElement.update(framework.dart:6595)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at 
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatelessElement.update(framework.dart:5162)
at ??RenderObjectElement.updateChildren(framework.dart:6093)
at ??MultiChildRenderObjectElement.update(framework.dart:6595)
at ??_ViewportElement.update(viewport.dart:236)
at 
at 
at 
at 
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at 
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at 
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatelessElement.update(framework.dart:5162)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatelessElement.update(framework.dart:5162)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatelessElement.update(framework.dart:5162)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatelessElement.update(framework.dart:5162)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??RenderObjectElement.updateChildren(framework.dart:6093)
at ??MultiChildRenderObjectElement.update(framework.dart:6595)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at 
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at 
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??ProxyElement.update(framework.dart:5417)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??StatefulElement.update(framework.dart:5274)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatelessElement.update(framework.dart:5162)
at ??ComponentElement.performRebuild(framework.dart:5111)
at ??StatefulElement.performRebuild(framework.dart:5251)
at ??BuildOwner.buildScope(framework.dart:2780)
at ??WidgetsBinding.drawFrame(binding.dart:903)
at ??RendererBinding._handlePersistentFrameCallback(binding.dart:358)
at ??RendererBinding._handlePersistentFrameCallback(binding.dart:357)
at ??SchedulerBinding._invokeFrameCallback(binding.dart:1289)
at ??SchedulerBinding.handleDrawFrame(binding.dart:1219)
at ??SchedulerBinding._handleDrawFrame(binding.dart:1077)
at ??SchedulerBinding._handleDrawFrame(binding.dart:1059)
at ??_invoke(hooks.dart:142)
at ??PlatformDispatcher._drawFrame(platform_dispatcher.dart:359)
at ??_drawFrame(hooks.dart:112)
at ??_drawFrame(hooks.dart:110)
```

**Comment by tvolkert** [comment]
Updating to P1 since the Google bug is listed as a blocker.

**Comment by goderbauer** [comment]
The crash appears to be happening while mounting (initializing) the Scrollable of a NestedScrollView, from the stack trace:

```
at ??ScrollableState.didChangeDependencies(scrollable.dart:599)
at ??StatefulElement._firstBuild(framework.dart:5237)
```

During mounting, a _NestedScrollPosition is created and it is going idle because there's no scrolling happening. As part of going idle, `_NestedScrollCoordinator.updateUserScrollDirection` is called to set the scroll direction to idle. And here is the first mystery: This method should short-circuit because `idle` is the initial scroll direction of the _NestedScrollCoordinator:

https://github.com/flutter/flutter/blob/0a63cd70f18e713bbfae4908d303daf76444d094/packages/flutter/lib/src/widgets/nested_scroll_view.dart#L628-L630

Since we are in the process of initializing the NestedScrollView it is surprising that its _NestedScrollCoordinator would have anything bug an idle scroll direction at this point. It's unclear to me why the scroll direction isn't still idle here...

But anyways, since the short circuit doesn't trigger, we continue and first fail to look up a MediaQuery. This is another surprise. Virtually any real-life app contains a media query injected by WidgetApp. This failure then causes us to look up a View, which fails hard and crashes:

https://github.com/flutter/flutter/blob/0a63cd70f18e713bbfae4908d303daf76444d094/packages/flutter/lib/src/widgets/scroll_position.dart#L244

This is another surprise because any app bootstrapped with `runApp` contains a View. Why is it not visible from this lookup?

Basically, there are two interesting questions to answer:

1) Why is the short-circuiting logic in _NestedScrollCoordinator.updateUserScrollDirection not triggering?
2) Why are MediaQuery and/or View (which are basically guaranteed to be there) not visible from the given context?

**Comment by goderbauer** [comment]
So far, I have been unsuccessful in repro'ing this. I cannot get the NestedScrollView into a state during initialization where the short-circuit logic doesn't trigger.

**Comment by goderbauer** [comment]
I have a repro for this now. Run the following with `flutter run --profile` and start scrolling. After five seconds the timer fires and restructures the widget tree, which crashes if a scroll is in progress:

```dart
import 'dart:async';

import 'package:flutter/material.dart';

void main() => runApp(const NestedScrollViewExampleApp());

class NestedScrollViewExampleApp extends StatelessWidget {
  const NestedScrollViewExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(body: NestedScrollViewExample()),
    );
  }
}

class NestedScrollViewExample extends StatefulWidget {
  const NestedScrollViewExample({super.key});

  @override
  State<NestedScrollViewExample> createState() => _NestedScrollViewExampleState();
}

class _NestedScrollViewExampleState extends State<NestedScrollViewExample> {
  bool nested = false;

  @override
  void initState() {
    super.initState();
    Timer(const Duration(seconds: 5), () {
      print('get ready');
      setState(() {
        nested = !nested;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    final Widget listView = ListView(
      children: const <Widget>[
        SizedBox(
          height: 1000,
        ),
      ],
    );
    return NestedScrollView(
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return <Widget>[
          SliverAppBar(
            title: const Text('Books'),
            pinned: true,
            expandedHeight: 150.0,
            forceElevated: innerBoxIsScrolled,
          ),
        ];
      },
      body: nested ? Container(child: listView) : listView,
    );
  }
}
```

Will clean this up into a nice regression test case shortly.

Quick analysis: When the timer fires, the widget subtree making up the inner scrollable is restructured, which causes the inner Scrollable to be disposed and a new one is initialized. This new Scrollable wants to goIdle when initialized, but since a Scroll is already in progress, `userScrollDirection` in _NestedScrollCoordinator.updateUserScrollDirection is not set to idle and that's why the short-circuit logic doesn't trigger. This answers open question 1 from above. We them proceed to look up MediaQuery/View and apparently the context we use for that is no longer valid, which - in release and profile mode - produces the TypeError we see in the crash. In debug mode, you get a nicer exception. The invalid context is our answer to question 2 and the reason why we cannot find the MediaQuery and/or View.

This leaves us with two new debugging questions:

1) Should the NestedScrollView abort a scroll-in-progress (and go idle) when the widget subtree making up the inner scrollable restructures? Or should the scroll remain in progress?
2) Why is the context used for lookup becoming invalid in this scenario?

## Workaround

You can give the inner scrolling widget a global key to make sure it is retained (and not disposed and reinitialized) when the the widget tree restructures (see example below) - or you can just avoid restructuring the widget tree like that.

<details>
<summary>Workaround with global key</summary>

```dart
import 'dart:async';

import 'package:flutter/material.dart';

void main() => runApp(const NestedScrollViewExampleApp());

class NestedScrollViewExampleApp extends StatelessWidget {
  const NestedScrollViewExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(body: NestedScrollViewExample()),
    );
  }
}

class NestedScrollViewExample extends StatefulWidget {
  const NestedScrollViewExample({super.key});

  @override
  State<NestedScrollViewExample> createState() => _NestedScrollViewExampleState();
}

class _NestedScrollViewExampleState extends State<NestedScrollViewExample> {
  bool nested = false;

  // WORKAROUND
  final GlobalKey _scrollableKey = GlobalKey();

  @override
  void initState() {
    super.initState();
    Timer(const Duration(seconds: 5), () {
      print('get ready');
      setState(() {
        nested = !nested;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    final Widget innerScrollable = ListView(
      key: _scrollableKey, // WORKAROUND
      children: const <Widget>[
        SizedBox(
          height: 1000,
        ),
      ],
    );
    return NestedScrollView(
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return <Widget>[
          SliverAppBar(
            title: const Text('Books'),
            pinned: true,
            expandedHeight: 150.0,
            forceElevated: innerBoxIsScrolled,
          ),
        ];
      },
      body: nested ? Container(child: innerScrollable) : innerScrollable,
    );
  }
}
```

</details>

**Comment by goderbauer** [comment]
Regression test for this issue:

```dart
  testWidgets('NestedScrollView does not crash when inner scrollable changes while scrolling', (WidgetTester tester) async {
    // Regression test for https://github.com/flutter/flutter/issues/126454.
    Widget buildApp({required bool nested}) {
      final Widget innerScrollable = ListView(
        children: const <Widget>[SizedBox(height: 1000)],
      );
      return MaterialApp(
        home: Scaffold(
          body: NestedScrollView(
            headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
              return <Widget>[
                SliverAppBar(
                  title: const Text('Books'),
                  pinned: true,
                  expandedHeight: 150.0,
                  forceElevated: innerBoxIsScrolled,
                ),
              ];
            },
            body: nested ? Container(child: innerScrollable) : innerScrollable,
          ),
        ),
      );
    }

    await tester.pumpWidget(buildApp(nested: false));

    // Start a scroll.
    final TestGesture scrollDrag = await tester.startGesture(tester.getCenter(find.byType(ListView)));
    await tester.pump();
    await scrollDrag.moveBy(const Offset(0, 50));
    await tester.pump();

    // Restructuring inner scrollable while scroll is in progress shouldn't crash.
    await tester.pumpWidget(buildApp(nested: true));
  });
```

**Comment by goderbauer** [comment]
This `goIdle` call here is questionable when the inner scrollable is replaced while a scroll is underway:

https://github.com/flutter/flutter/blob/0a63cd70f18e713bbfae4908d303daf76444d094/packages/flutter/lib/src/widgets/nested_scroll_view.dart#L1183-L1185

We are only reporting idle for a split second because the scroll is still underway. It's a strange glitch. I wonder, if the new nested scrollable should somehow absorb the scroll position of the old nested scrollable to avoid that glitch? cc @Piinks Also, do you remember why the _NestedScrollPosition has to explicitly go idle when instantiated? The regular ScrollPosition doesn't do this...

**Comment by goderbauer** [comment]
Proposed fix: https://github.com/flutter/flutter/pull/126535
